src/sql/workbench/api/node/mainThreadNotebookDocumentsAndEditors.ts

```diff
@@ -0,0 +1,367 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+'use strict';
+
+import * as sqlops from 'sqlops';
+import { extHostNamedCustomer } from 'vs/workbench/api/electron-browser/extHostCustomers';
+import { Disposable } from 'vs/base/common/lifecycle';
+
+import URI, { UriComponents } from 'vs/base/common/uri';
+import { IExtHostContext } from 'vs/workbench/api/node/extHost.protocol';
+import { IEditorService } from 'vs/workbench/services/editor/common/editorService';
+import { IEditorGroupsService } from 'vs/workbench/services/group/common/editorGroupsService';
+import { IInstantiationService } from 'vs/platform/instantiation/common/instantiation';
+import { ITextEditorOptions } from 'vs/platform/editor/common/editor';
+import { viewColumnToEditorGroup } from 'vs/workbench/api/shared/editor';
+
+import {
+	SqlMainContext, MainThreadNotebookDocumentsAndEditorsShape, SqlExtHostContext, ExtHostNotebookDocumentsAndEditorsShape,
+	INotebookDocumentsAndEditorsDelta, INotebookEditorAddData, INotebookShowOptions, INotebookModelAddedData
+} from 'sql/workbench/api/node/sqlExtHost.protocol';
+import { NotebookInputModel, NotebookInput } from 'sql/parts/notebook/notebookInput';
+import { INotebookService, INotebookEditor } from 'sql/services/notebook/notebookService';
+import { TPromise } from 'vs/base/common/winjs.base';
+
+class MainThreadNotebookEditor extends Disposable {
+
+	constructor(public readonly editor: INotebookEditor) {
+		super();
+	}
+
+	public get uri(): URI {
+		return this.editor.notebookParams.notebookUri;
+	}
+
+	public get id(): string {
+		return this.editor.id;
+	}
+
+	public get isDirty(): boolean {
+		return this.editor.isDirty();
+	}
+
+	public get providerId(): string {
+		return this.editor.notebookParams.providerId;
+	}
+
+	public save(): Thenable<boolean> {
+		return this.editor.save();
+	}
+
+	public matches(input: NotebookInput): boolean {
+		if (!input) {
+			return false;
+		}
+		return input === this.editor.notebookParams.input;
+	}
+}
+
+function wait(timeMs: number): Promise<void> {
+	return new Promise(resolve => setTimeout(resolve, timeMs));
+}
+
+
+namespace mapset {
```

Comment:
`mapset` and `delta` namespaces are ported from existing vscode logic. I copied instead of changing the file to export these to avoid new SQL CARBON EDIT tags